# Velocity adding in special relativity

## What happens when you add two speeds close to light speed?

Two objects are passing each other, each travelling at 99.99% light speed relative to a stationary object. What velocity do they measure for each other? More than light speed?

No, heres why.

## In everyday life

In everyday life, we can just add velocities like this:

\begin{equation}
v_{\text{total}} = v_1 + v_2
\end{equation}

So if an object travelling at 100 metres per second passes another object going at 100 m/s in the opposite direction, they measure each other's speed as just:

\begin{equation}
100 + 100 = 200
\end{equation}

But this common sense approach is only an approximation, and breaks down as speeds get closer to $c$ (light speed).

## Relativistic velocity addition

The correct way to add any velocities is like this:

\begin{equation}
v_{\text{total}} = \frac{v_1 + v_2}{1 + \frac{v_1 v_2}{c^2}}
\end{equation}

The difference is microscopic for everyday speeds, so no one notices. But as you get closer to light speed it becomes more significant.

The new equation is a bit more complicated, so let's write some Python code to help us in this Jupyter Notebook.

*You can download this notebook, and experiment with it on your PC or device. Try out your own velocity additions if you like!*

In [7]:
import math

c = 299_792_458.0  # speed of light, metres per second
c_squared = c**2  # c squared


def add_vel(v1: float, v2: float) -> float:
    if v1 >= c or v2 >= c:
        return math.nan
    return (v1 + v2) / (1.0 + (v1 * v2) / c_squared)


# test it
print(f"100 + 100 = {add_vel(100, 100)}")

100 + 100 = 199.99999999997775


Almost the same answer, to within 10 decimal places. Let's try it with some faster speeds.

In [8]:
def calculate(v: float, round: bool = True) -> None:
    simple = v + v
    relativistic = add_vel(v, v)
    diff = 1.0 - relativistic / simple  # difference as a percentage
    if round:
        print(f"{v:,} m/s + {v:,} m/s = {relativistic:,.1f} (discrepancy {diff:%})")
    else:
        print(f"{v:,} m/s + {v:,} m/s = {relativistic:,} (discrepancy {diff:%})")


# change these, but make sure they are less than light speed
calculate(1_000, False)
calculate(100_000, False)
calculate(1_000_000)
calculate(10_000_000)
calculate(100_000_000)
calculate(200_000_000)

1,000 m/s + 1,000 m/s = 1,999.9999999777472 (discrepancy 0.000000%)
100,000 m/s + 100,000 m/s = 199,999.97774700134 (discrepancy 0.000011%)
1,000,000 m/s + 1,000,000 m/s = 1,999,977.7 (discrepancy 0.001113%)
10,000,000 m/s + 10,000,000 m/s = 19,977,771.7 (discrepancy 0.111141%)
100,000,000 m/s + 100,000,000 m/s = 179,975,072.5 (discrepancy 10.012464%)
200,000,000 m/s + 200,000,000 m/s = 276,805,111.1 (discrepancy 30.798722%)


## Effect gets proportionally larger

The difference between simple and relativistic addition gets larger as speeds increase. At 2 km/s it is almost zero, but at 400,000 km/s it has shot up to over 30%.

For faster speeds, let's switch to using fractions of $c$ (light speed) instead of m/s. Everything else is the same:

In [9]:
def add_vel_c(v1: float, v2: float) -> float:
    # two velocities as fraction of c
    if v1 >= 1.0 or v2 >= 1.0:
        return math.nan
    return (v1 + v2) / (1.0 + (v1 * v2))


def calculate_c(v1: float) -> None:
    simple = v1 + v1
    relativistic = add_vel_c(v1, v1)
    diff = 1.0 - relativistic / simple  # difference as a percentage
    print(f"{v1:,}c + {v1:,}c = {relativistic:,} (discrepancy {diff:.1%})")


# change these, but make sure they are less than light speed, 1.0
calculate_c(0.1)
calculate_c(0.2)
calculate_c(0.3)
calculate_c(0.4)
calculate_c(0.5)
calculate_c(0.7)
calculate_c(0.9)
calculate_c(0.95)
calculate_c(0.9999)
calculate_c(0.999999)
# calculate_c(1.0)

0.1c + 0.1c = 0.19801980198019803 (discrepancy 1.0%)
0.2c + 0.2c = 0.38461538461538464 (discrepancy 3.8%)
0.3c + 0.3c = 0.5504587155963302 (discrepancy 8.3%)
0.4c + 0.4c = 0.689655172413793 (discrepancy 13.8%)
0.5c + 0.5c = 0.8 (discrepancy 20.0%)
0.7c + 0.7c = 0.9395973154362416 (discrepancy 32.9%)
0.9c + 0.9c = 0.994475138121547 (discrepancy 44.8%)
0.95c + 0.95c = 0.9986859395532195 (discrepancy 47.4%)
0.9999c + 0.9999c = 0.9999999949995 (discrepancy 50.0%)
0.999999c + 0.999999c = 0.9999999999995001 (discrepancy 50.0%)


Notice how the discrepancy grows up to 50%, and the added velocities never exceed $c$

# So the answer to our question

99.99% c + 99.99% c = 99.99999949995% c

How about that! Regardless of your velocities, you will never measure a speed greater than $c$.

## More info

For more on relativistic velocities, check out 

https://en.wikipedia.org/wiki/Velocity-addition_formula#Special_relativity

[And my Python relativity tools](https://github.com/lookbusy1344/Relativity/blob/main/Python/README.md)